### PR TITLE
Prevent duplicate kernels booting and ensure they're cleaned up

### PR DESCRIPTION
**Pull request by ndench**
After upgrading to Symfony 4.4 we received a massive memory blow out when running tests, taking our memory using from `300MB` to `9GB`. I initially reported an issue to Symfony (symfony/symfony#35164) and after many months and a lot of debugging I identified symfony/symfony#34078 as the pull request which make the memory leak apparent.

There are a few smaller issues which come together to create the memory leak:
* As of Symfony 4.4+ they no longer reset the test container after tests, instead only resetting the real one
* Symfony KernelTestCase boots and keeps track of it's own kernel
* LiipTestFixtures creates and keeps track of another kernel, separate to Symfony
* Symfony WebTestCase::createClient() will always boot a kernel and booting a kernel before calling this function is deprecated in Symfony 4.4 (symfony/symfony#32056)
* Loading fixtures with LiipTestFixtures boots it's own kernel, regardless of whether it's called before or after WebTestCase::createClient()

To fix this we need to ensure only the kernel tracked by Symfony KernelTestCase is booted, because it will be automatically cleaned up by Symfony. If we want to run multiple kernels, we need to ensure previous ones are shut down before booting new ones (symfony/symfony#34507).

I had a quick look through old issues in LiipTestFixturesBundle and LiipFunctionalTestBundle and I believe that liip/LiipFunctionalTestBundle#152 and #12 will be fixed by this PR.

I'm unsure how to test this change (is testing necessary here?), if someone can point me in the right direction that would be appreciated.

### CODE REVIEW

File: src/Test/FixturesTrait.php [hunk 1]
```diff
@@ -45,15 +45,15 @@ protected function getContainer(): ContainerInterface
...
-                $this->containers[$environment] = $container;
+
+            // Check that the kernel has not been booted separately (eg. with static::createClient())
+            $kernel = static::$kernel;
+            if (null === $kernel) {
+                $kernel = $this->bootKernel($options);
             }
+
+            // bootKernel() has logic to ensure static::$container is the correct container
+            $this->containers[$environment] = static::$container;
```
Comment (ndench) [inline src/Test/FixturesTrait.php:56]: Looks like the `static::$container` property was introduced in Symfony 4, which is causing the build to fail for Symfony 3.4. Should I undo this change?
```suggestion
            $container = $kernel->getContainer();
            if ($container->has('test.service_container')) {
                $this->containers[$environment] = $container->get('test.service_container');
            } else {
                $this->containers[$environment] = $container;
            }
```

Comment (Tobion) [reply src/Test/FixturesTrait.php:56]: How about `$this->containers[$environment] = static::$container ?? $kernel->getContainer();`

File: src/Test/FixturesTrait.php [hunk 2]
```diff
@@ -45,15 +45,15 @@ protected function getContainer(): ContainerInterface
...
-
-            $container = $kernel->getContainer();
-            if ($container->has('test.service_container')) {
-                $this->containers[$environment] = $container->get('test.service_container');
-            } else {
-                $this->containers[$environment] = $container;
+
+            // Check that the kernel has not been booted separately (eg. with static::createClient())
+            $kernel = static::$kernel;
+            if (null === $kernel) {
```
Comment (Tobion) [inline src/Test/FixturesTrait.php:51]: you don't need a $kernel variable? `if (null === static::$kernel) { $this->bootKernel($options); }`

**Comment by Tobion** [comment]
> Symfony WebTestCase::createClient() will always boot a kernel and booting a kernel before calling this function is deprecated in Symfony 4.4 (symfony/symfony#32056)

This bundles documentation still seems to be affected by this deprecation:
It documentations something like
```
$this->loadFixtures();
$client = $this->createClient();
```

in https://github.com/liip/LiipTestFixturesBundle/blob/master/doc/database.md and https://github.com/liip/LiipTestFixturesBundle/blob/master/doc/examples.md.
This would trigger the deprecation wouldn't it?

File: src/Test/FixturesTrait.php (hunk 1 above)
Comment (ndench) [reply src/Test/FixturesTrait.php:56]: This still won't work on Symfony 3.4 I don't think. Because it will get the "real" container instead of the test one. eg. this is how KernelTestCase does it in Symfony 4:

```php
self::$kernelContainer = $container = static::$kernel->getContainer();
static::$container = $container->has('test.service_container') ? $container->get('test.service_container') : $container;
```

I think I'll just revert the change.

**Comment by ndench** [comment]
Thanks for the review @Tobion, good catch on the documentation. I've implemented your fixes.

**Comment by ndench** [comment]
Looks like the build failed again, but for a different reason. Seems to be in can't install a dependency? I'm not sure how this PR affects that though.

**Comment by Tobion** [comment]
Lowest dependency tests fail with
>Error: Call to a member function has() on null
>/home/travis/build/liip/LiipTestFixturesBundle/src/Test/FixturesTrait.php:55

**Comment by ndench** [comment]
Ah, I was looking at the build that's allowed to fail.

Turns out the KernelTestCase still keeps `static::$kernel` around after it's shut the kernel down so we need to check if the kernel still has a container as well.

**Comment by alexislefebvre** [comment]
Thanks a lot @ndench and @Tobion! Here is the last version, which contain this PR: https://github.com/liip/LiipTestFixturesBundle/releases/tag/1.8.0

**Comment by bendavies** [comment]
so how are we meant to use [loginAs](https://github.com/liip/LiipFunctionalTestBundle/blob/master/doc/logged.md#logging-in-a-user-using-webtestcaseloginas) with this change?

`loginAs` requires the user from fixtures, but now you need to create the client before loading the fixtures.
So we have a chicken egg problem.
This no longer works:
```php
$fixtures = $this->loadFixtures(array(
    'AppBundle\DataFixtures\ORM\LoadUserData'
))->getReferenceRepository();

$this->loginAs($fixtures->getReference('account-alpha'), 'main');
$client = $this->makeClient();
```

**Comment by ndench** [comment]
Ah, good catch @bendavies. I wasn't aware that this was even a feature!

Indeed in it looks as if `$this->makeClient()` follows the [Symfony recommendations](https://symfony.com/doc/current/testing/http_authentication.html) on how to authenticate a client. In our codebase we created a helper function which does just this before we started using the FunctionalTestBundle and continued using it afterwards. So I completely missed this.

If you're lucky enough to be running Symfony 5.1, then this is solved for you by the introduction of [`$client->loginUser()`](https://symfony.com/blog/new-in-symfony-5-1-simpler-login-in-tests), though I'm sure there are a lot of people on older versions (as am I).

We can't really revert this change, since Symfony has deprecated calling `static::createClient()` after the kernel has already been booted. However I think we could move the login logic to a new function, such as `$this->loginWithClient($client, $user, $firewall)`, which would allow you do:

```php
$client = $this->makeClient();

$fixtures = $this->loadFixtures(array(
    'AppBundle\DataFixtures\ORM\LoadUserData'
))->getReferenceRepository();

$this->loginWithClient($client, $fixtures->getReference('account-alpha'), 'main');
```

What do you think?

**Comment by bendavies** [comment]
Thanks for the detailed reply! A new function sounds good to me!

**Comment by PapyDanone** [comment]
Yep this broke my entire functional test suite, because of the use of `loginAs` 😭  

Out of curiosity why the need of a new function, couldn't we leverage the new `$client->loginUser()` from Symfony as noted by @ndench ?

**Comment by ndench** [comment]
You can use the new Symfony function, but it's only available on Symfony 5.
So if you're on an older version then you'll need a new function the manage
it.

**Comment by PapyDanone** [comment]
@ndench looks like a [5.1 feature indeed](https://symfony.com/blog/new-in-symfony-5-1-simpler-login-in-tests).

I did fix some of my tests using an intermediate login method as you suggested. 

Now other tests using the entity manager's `refresh()` method seem to be busted as well. I can't figure out whether it's because of the changes in this PR.

```php
$this->client = $this->createClient();
$this->fixtures = $this->loadFixtures([...])->getReferenceRepository();
$this->em = $this->client->getContainer()->get('doctrine')->getManager();

$this->user = $this->createApiUser(); // throw away user for underlying API
$this->em->persist($this->user);
$this->em->flush();
$this->loginAsFix($this->client, $this->user, 'main');
// [...] API modifies database user
$this->em->refresh($this->user);

/* ERROR: Doctrine\ORM\ORMInvalidArgumentException:
Entity retry Investor is not managed. An entity is managed if its fetched 
from the database or registered as new through EntityManager#persist */
```
Any

**Comment by ndench** [comment]
@PapyDanone sorry, I must have missed this notifications.

One thing I noticed after enforcing a single kernel is that I no longer need to refresh the entity manager in tests. Before this change I used to:

1. Load fixtures
2. Get required entities from the fixtures
3. Create client
4. Run request
5. Refresh entities
6. Assert entities are updated as expected

But after the change, I can leave out step 5 and all the tests pass.

My only assumption here is that the kernel that was booted to load the fixtures is a different kernel to the one the request was executed under (because KernelTestCase::createClient always boots another kernel). This means that the entities are loaded from the database using a different instance of Doctrine and thus a different cache.

But now that only one kernel is being used, I think that the request is modifying the same instance of the objects created by fixtures. Which means tests don't need to get the changes from the database anymore.

I haven't verified any of this, but that's my gut feel.

**Comment by ndench** [comment]
I've created Liip/LiipFunctionalTestBundle#567 to add the `loginClient` functionality.

**Comment by PapyDanone** [comment]
Thank you for your detailed response @ndench, much appreciated. And sorry for the late reply, I only got the chance to get back to this. 

I started using `loginClient` and I still have the issue of the entity manager losing the reference to the entity after the original `persist`.

After some debugging it turns out it happens after a redirect + follow. 

Using `var_dump($this->em->contains($this->user));`, it returns `true` before the redirect and `false` after, so it seems like the issue is session related.

**Comment by rairlie** [comment]
@ndench 
> My only assumption here is that the kernel that was booted to load the fixtures is a different kernel to the one the request was executed under (because KernelTestCase::createClient always boots another kernel). This means that the entities are loaded from the database using a different instance of Doctrine and thus a different cache.
> 
> But now that only one kernel is being used, I think that the request is modifying the same instance of the objects created by fixtures. Which means tests don't need to get the changes from the database anymore.

This causes some test failures for us when updating to 1.8.0. Since the application under test uses the same kernel as the test setup, we see a difference in behaviour. 

In one example, we have some service code that looks like this:

```
public function deleteUser(int $id): bool
{
    $user = $this->userRepo->find($id);

    if (!$user) {
        return false;
    }
```

And a functional test like this:
```
public function testDeleteAlreadyDeletedUser(): void
{
    $userId = $this->fixtureRepo->getReference('deleted_user')->getId();

    $this->assertFalse($this->userService->deleteUser($userId));
}
```

What should happen is `$this->userRepo->find($id)` returns null, because we have a doctrine filter that modifies the `find()` query to hide soft-deleted entities. But what actually happens is the entity is returned from cache without hitting the DB, because its in the UnitOfWork shared with the test kernel EntityManager.

In principal this feels wrong - I would not expect the fixture setup to change the behaviour of the code being tested. We can manually reset the UnitOfWork before the test, but I'm not sure that's the right approach.

If I downgrade to 1.7.0 I get the high memory usage problem (we are using symfony 4.4), however if I apply this change: https://github.com/symfony/symfony/issues/35164#issuecomment-616585230 then everything works great.

**Comment by ndench** [comment]
@rairlie have tried getting the EntityManager or UserRepository and calling `clear()` before the service function call? That will empty the cache 

I'm not sure why the change in that comment makes a difference though. It looks very similar to this PR?

**Comment by rairlie** [comment]
@ndench 

The difference is this PR uses the existing `static::$kernel`, while the other change creates its own `static::$localKernel` for loading fixtures.

So with the PR the flow is:
```
    $this->makeClient();
    // -> Liip\FunctionalTestBundle\Test::makeClient()
    // -> Symfony\Bundle\FrameworkBundle\Test\WebTestCase::createClient()
    // -> Symfony\Bundle\FrameworkBundle\Test\KernelTestCase::bootKernel()
    // static::$kernel = the booted kernel

    $this->loadFixtures();
    // Liip\TestFixturesBundle\Test\FixturesTrait::loadFixtures()
    // Liip\TestFixturesBundle\Test\FixturesTrait::getContainer()
    // kernel already booted - uses existing static::$kernel
    // fixtures are loaded using the static::$kernel

    $serviceToTest = static::$container->get('my_service');
    // Service being tested uses same static::$kernel as the fixture setup, so
    // is wired to same entity manager, has cached entities
```

With the change in https://github.com/symfony/symfony/issues/35164#issuecomment-616585230 we see:
```
    $this->makeClient();
    // same as before - static::$kernel = the booted kernel

    $this->loadFixtures();
    // Liip\TestFixturesBundle\Test\FixturesTrait::loadFixtures()
    // Tests\MyBaseWebTestCase::getContainer()
    // creates a new static::$localKernel
    // fixtures are loaded using the static::$localKernel

    $serviceToTest = static::$container->get('my_service');
    // Service being tested uses static::$kernel -- different kernel from
    // the fixture setup, so starts with a clean entity manager
```

I would not expect loading fixtures to populate the EM cache, such that I need to clear the cache before my test. If it is intended behaviour it seems to be a breaking change, as we didn't do it previously. Or is there maybe something I'm missing?

**Comment by ndench** [comment]
Ah yeah, I see the difference here @rairlie. In my opinion we shouldn't have multiple kernels being booted, it creates memory leaks and the Symfony framework does not expect it (Symfony 4.4 explicitly deprecates and tries to prevent it from happening, which is part of what caused this PR in the first place).

That being said, I understand that it's pretty damn annoying that your tests used to work and they don't anymore. I think that maybe we should have incremented the major version for this since it did break something. However, it was really a breaking change in Symfony that caused it. 

I'm unsure the best way to go about this moving forward. But I think the first step would be to create a new issue to move discussions away from a closed PR.
